### PR TITLE
Improve S3776: Exclude complexity of JSX short-circuits

### PR DESCRIPTION
**Pull request by francoismora**
Fixes SonarSource/SonarJS#3289

### CODE REVIEW

File: src/rules/cognitive-complexity.ts [hunk 1]
```diff
@@ -113,6 +113,9 @@ const rule: TSESLint.RuleModule<string, (number | 'metric' | 'sonar-runtime')[]>
     /** Stack of enclosing functions */
     const enclosingFunctions: TSESTree.FunctionLike[] = [];
 
+    /** Indicator if inside a JSX expression */
+    let jsx = false;
```
Comment (ilia-kebets-sonarsource) [inline src/rules/cognitive-complexity.ts:117]: `isJsx`?

Comment (francoismora) [reply src/rules/cognitive-complexity.ts]: Renamed to jsxLevel as this should be a number.

**Comment by francoismora** [comment]
As discussed, the rule checks that the short-circuit logical expression doesn't have a ternary operator at its right-hand side before deciding to exclude it from the cognitive complexity computation.
Also, the PR fixes an issue in the secondary locations of logical operators.

File: src/rules/cognitive-complexity.ts [hunk 2]
```diff
@@ -339,20 +338,47 @@ const rule: TSESLint.RuleModule<string, (number | 'metric' | 'sonar-runtime')[]>
...
+        return;
+      }
+
       if (!consideredLogicalExpressions.has(logicalExpression)) {
         const flattenedLogicalExpressions = flattenLogicalExpression(logicalExpression);
 
         let previous: TSESTree.LogicalExpression | undefined;
         for (const current of flattenedLogicalExpressions) {
           if (!previous || previous.operator !== current.operator) {
-            const operatorTokenLoc = getFirstTokenAfter(logicalExpression.left, context)!.loc;
```
Comment (ilia-kebets-sonarsource) [inline src/rules/cognitive-complexity.ts:348]: why this change?

Comment (francoismora) [reply src/rules/cognitive-complexity.ts:348]: That's the issue I mention in my comment above. All the locations raised by this function are the same instead of pointing to the flattened logical operators.

Comment (francoismora) [reply src/rules/cognitive-complexity.ts:348]: FYI, I added a test case for this fix in the last commit.

File: src/rules/cognitive-complexity.ts [hunk 3]
```diff
@@ -339,20 +339,47 @@ const rule: TSESLint.RuleModule<string, (number | 'metric' | 'sonar-runtime')[]>
...
+        logicalExpression.parent?.type === 'JSXExpressionContainer' &&
+        isJsxShortCircuitSubNode(logicalExpression, logicalExpression.left) &&
+        isJsxShortCircuitSubNode(logicalExpression, logicalExpression.right);
+      if (isShortCircuit) {
+        consideredLogicalExpressions.add(logicalExpression);
+      }
+      return isShortCircuit;
+    }
+
+    function isJsxShortCircuitSubNode(root: TSESTree.LogicalExpression, node: TSESTree.Node) {
```
Comment (ilia-kebets-sonarsource) [inline src/rules/cognitive-complexity.ts:371]: So the maximum of allowed binary operators is 2?

Comment (francoismora) [reply src/rules/cognitive-complexity.ts]: As discussed, the last commit adds recursion logic to the short-circuit check.

File: src/rules/cognitive-complexity.ts [hunk 4]
```diff
@@ -357,26 +359,32 @@ const rule: TSESLint.RuleModule<string, (number | 'metric' | 'sonar-runtime')[]>
...
 
-    function isJsxShortCircuitSubNode(root: TSESTree.LogicalExpression, node: TSESTree.Node) {
-      if (node.type === 'LogicalExpression') {
-        const isShortCircuit = node.operator === root.operator;
-        if (isShortCircuit) {
-          consideredLogicalExpressions.add(node);
-        }
-        return isShortCircuit;
+    function getJsxShortCircuitSubNodes(
+      root: TSESTree.LogicalExpression,
```
Comment (ilia-kebets-sonarsource) [inline src/rules/cognitive-complexity.ts:371]: maybe pass only the operator?

File: src/rules/cognitive-complexity.ts [hunk 5]
```diff
@@ -339,7 +339,9 @@ const rule: TSESLint.RuleModule<string, (number | 'metric' | 'sonar-runtime')[]>
     }
 
     function visitLogicalExpression(logicalExpression: TSESTree.LogicalExpression) {
-      if (isJsxShortCircuitNode(logicalExpression)) {
+      const jsxShortCircuitNodes = getJsxShortCircuitNodes(logicalExpression);
+      if (jsxShortCircuitNodes) {
```
Comment (ilia-kebets-sonarsource) [inline src/rules/cognitive-complexity.ts:343]: `jsxShortCircuitNodes.length > 0`?

**Review by ilia-kebets-sonarsource — APPROVED** [review]
2 comments, otherwise LGTM!

File: src/rules/cognitive-complexity.ts (hunk 4 above)
Comment (francoismora) [reply src/rules/cognitive-complexity.ts:371]: Not sure about this as passing simply the node hides the implementation details that the check uses only the operator. Also, I don't think it improves anything in terms of performance, does it? But I guess it's simply a matter of preference.

File: src/rules/cognitive-complexity.ts (hunk 5 above)
Comment (francoismora) [reply src/rules/cognitive-complexity.ts:343]: It can also be `null` and if not its length can't be `0`, so it could be 
`jsxShortCircuitNodes != null`
so that's why the truthy check seemed sufficient. But I will change it to be clearer.
